### ISSUE TITLE
idea: hidden fields

### ISSUE THREAD

**Issue by colin-kiegel** [colin-kiegel/rust-derive-builder#15]
# use case

User wants to expose custom setters instead, which are more complex, do additional checks, etc.
# possible syntax

field annotation:
- hide: `#[builder(hide=true)]`
- unhide `#[builder(hide=false)]`

shorthand
- hide: `#[builder(hide)]`
- unhide: `#[builder]`

# Example

``` rust
#[derive(Builder)]
struct Foo {
    #[builder(hide)]
    foo: Vec<i32>
}
```
# alternatives
- hide
- ignore
- omit
- skip
- no_setter (i.e. `#[builder(no_setter)]` instead of `#[builder(setter(skip))]` ...)
- ...

**Comment by dpc** [comment]
It seems to me lack of this prevents me from using this (BTW. really nice) crate.

The struct that I want to build has way more fields than the ones I want to expose in the builder crate.

**Comment by colin-kiegel** [comment]
@dpc thx for your feedback. Would all hidden fields implement the `Default` trait, or do you want to annotate a field with a custom default specific only to the build method (e.g. `#[builder(hide, default="vec![1,2,3]")])`?

**Comment by dpc** [comment]
The thing is: I'd actually want to annotate/describe the `struct MyStructBuilder` itself, not `MyStruct`. The `struct MyStruct` is fancy, contains data like channels, `Arc`s of weird things, etc. 

What I'd want is:

```
#[derive(Builder)]]
struct MySructBuilder {
   field1 : int, 
   field2 : bool,
   ...
}
```

that would generate a code all the setters and:

```
impl MyStructBuilder {
    fn build(self) -> MyStruct {
          MyStruct::from(self)  // or just self.into() ?
    }
}
```

or something like that.

This way I can describe the builder without impacting how `MyStruct` is supposed to work and look like. Also, if I wanted to I could have multiple different builders for my crate (with some argument support) etc.

**Comment by colin-kiegel** [comment]
@dpc Interesting. The bottomline is that you would write `impl From<MyStructBuilder> for MyStruct` _yourself_, right?

I have the feeling this might actually be a use case for the [accessors](https://github.com/emk/accessors) crate. What if you use that to use `#[derive(setters)]]` instead
```rust
#[derive(setters)]]
struct MySructBuilder {
   field1 : int, 
   field2 : bool,
   ...
}
```

Then impl the `From` trait and treat `.into()` as your build method?

Would that solve your problem, or would you miss anything that only a dedicated `builder` crate could give you - but not an `accessors` crate?

**Comment by colin-kiegel** [comment]
PS: That being said, we still have the plan to hide individual setters, have defaults for fields and possibly omit the build method completely, or modify it. But I am not yet sure if we want to support the user to write `MyStructBuilder` himself - that's why I am asking. :-)

**Comment by dpc** [comment]
In my example I would `impl From<MyStructBuilder> for MyStruct`, but to be precise, the bottom line is: What `MyStructBuilder` holds is not what `MyStruct` will hold. `MyStruct` will just use builder fields as information on how to build itself.

Would `accessors` give me the builder chaining pattern ( taking and returning `&mut self`/`self`)? If so, I guess it would do just fine.

**Comment by colin-kiegel** [comment]
@dpc: Hm, `accessors` [does not](https://github.com/emk/accessors/blob/master/src/lib.rs#L96-L106) support chainint _yet_. But it would not be too complicated to send a PR and improve that - you only need to change 4 lines of code. And I believe it makes more sense in that crate, too - because all you need is to derive setters of some sort. :-)

Yes this crate only supports 1:1 mapping between builder and product, but we will most certainly relax that.

**Comment by dpc** [comment]
All right. I'll look into it. Thanks!

**Comment by colin-kiegel** [comment]
I thought we could have a very compact shorthand notation like this
- hide: `#[builder(hide)]`
- unhide: `#[builder]`

But this might collide with other extensions like [custom defaults](https://github.com/colin-kiegel/rust-derive-builder/issues/3), e.g.
```rust
#[derive(Builder)]
#[builder(hide))]
struct SkipSetterShorthand {
    hidden: u32,
    #[builder]
    visible_override_shorthand: u32,
    #[builder(default="42")] // <-- this would 'unhide' the annotated field 
    hidden_custom_default: u32,
}
```

So either we need to drop the shorthand notation, or put into a `setter` namespace. Now I am leaning towards
shorthand notation:
- hide: `#[builder(setter(skip)]`
- unhide: `#[builder(setter)]` (or any `#[builder(setter(..))]` that does not contain `skip`).

explicit notation:
- hide: `#[builder(setter(skip=true))]`
- unhide `#[builder(setter(skip=false))]`

**Comment by colin-kiegel** [comment]
I have a PR almost ready - only need to write some docs :-)

**Comment by colin-kiegel** [comment]
published with [v0.4.0](https://github.com/colin-kiegel/rust-derive-builder/releases/tag/v0.4.0)
